### PR TITLE
Fix view_func replay in no-grad mode

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83686
* __->__ #83872

Fixes https://github.com/pytorch/pytorch/issues/83828

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -1260,6 +1260,13 @@ def backward(ctx, grad_b):
 
         TestFn.apply(b).sum().backward()
 
+    def test_first_grad_fn_access_in_no_grad_mode(self):
+        a = torch.tensor([1 + 1j], requires_grad=True).clone()
+        v = a.real
+        a.add_(1)
+        with torch.autograd.grad_mode.no_grad():
+            v.grad_fn
```
Comment (albanD) [inline test/test_autograd.py:1268]: Don't you need to try and use it for the error to happen?

**Review by albanD — APPROVED** [review]
Oupsy

File: test/test_autograd.py (hunk 1 above)
Comment (soulitzer) [reply test/test_autograd.py:1268]: Actually refreshing grad_fn also tries to update the cpp hooks, which checks that the grad_fn passed in is not nullptr

Comment (albanD) [reply test/test_autograd.py:1268]: Ok!

**Pull request closed without merging**